### PR TITLE
uac: Fix 2 spaces when replacing empty display name

### PR DESCRIPTION
**Pull request by xkaraman**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [x] PR should be backported to stable branches
- [x] Tested changes locally
- [x] Related to issue #4582

#### Description

PR aims to fix the 2 spaced related to the references issued. 

After reading the rfc of what is allow for the header fields, i tried to consider most of the cases and output the recommended form of `Header: header_value`, one space after colon. Also one space after the Display name.

### CODE REVIEW

File: src/modules/uac/replace.c [hunk 1]
```diff
@@ -343,6 +343,9 @@ int replace_uri(struct sip_msg *msg, str *display, str *uri,
 				pkg_free(buf.s);
 				goto error;
 			}
+		} else {
+			/* deal with extra whitespace when the display is removed */
+			l = del_lump(msg, body->display.s - msg->buf - 1, 1, 0);
```
Comment (miconda) [inline src/modules/uac/replace.c]: It is not mandatory to have the space between the `:` after the header name and the display name. If such space is not there, I think the above line is going to delete the `:`, resulting in a broken header. A check to be sure there is a white space should be done.

Looking at the code, maybe deleting the white space should combined with the lump deleting the old display name, when the new display name is empty.

Comment (xkaraman) [reply src/modules/uac/replace.c]: Hmm ok didn't consider that case. will adopt.

**Comment by xkaraman** [comment]
This got a bit more than initially thought due to many considerations like multiple spaces before and after, no spaces, with/out display name etc. 

I used multiple messages to test which can be found here:

<details> 
<summary>  test_message.sh </summary>

```
SIP_SERVER="kamailio.server.com"
SIP_PORT="5060"

send_invite() {
	printf "%b" "$1" | timeout 0.3 nc -u -4 "$SIP_SERVER" "$SIP_PORT"
}

# Valid case: From with display name and angle brackets, To with display name and angle brackets

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: Alice Smith <sip:alice@there.com>;tag=1001\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf95\r\n\
Max-Forwards: 70\r\n\
To: Bob Jones <sip:bob@example.com>\r\n\
Call-ID: test-case-001\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Valid case: From with angle brackets no display name, To with angle brackets no display name

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: <sip:alice@there.com>;tag=1002\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf91\r\n\
Max-Forwards: 70\r\n\
To: <sip:bob@example.com>\r\n\
Call-ID: test-case-002\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Valid case: From without angle brackets, To without angle brackets

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: sip:alice@there.com;tag=1003\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf92\r\n\
Max-Forwards: 70\r\n\
To: sip:bob@example.com\r\n\
Call-ID: test-case-003\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Invalid case: From with display name no angle brackets, To with display name no angle brackets

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: Alice sip:alice@there.com;tag=1004\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf93\r\n\
Max-Forwards: 70\r\n\
To: Bob sip:bob@example.com\r\n\
Call-ID: test-case-004\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Invalid case: From missing closing angle bracket

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: Alice <sip:alice@there.com;tag=1005\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf95\r\n\
Max-Forwards: 70\r\n\
To: <sip:bob@example.com>\r\n\
Call-ID: test-case-005\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Valid case syntax: To with malformed URI

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: Alice <sip:alice@there.com>;tag=1006\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf96\r\n\
Max-Forwards: 70\r\n\
To: <sip:bob@>\r\n\
Call-ID: test-case-006\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# valid case: From missing tag parameter

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: <sip:alice@there.com>\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf97\r\n\
Max-Forwards: 70\r\n\
To: <sip:bob@example.com>\r\n\
Call-ID: test-case-007\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Edge case: From with display name containing special characters, with angle brackets

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: \"Alice Smith, Support\" <sip:alice@there.com>;tag=1008\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf98\r\n\
Max-Forwards: 70\r\n\
To: \"Bob Jones, Sales\" <sip:bob@example.com>\r\n\
Call-ID: test-case-008\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Valid case: From and To with port numbers in URI

INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: Alice Smith <sip:alice@there.com;port=5050>;tag=1001\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf99\r\n\
Max-Forwards: 70\r\n\
To: Bob Jones <sip:bob@example.com>\r\n\
Call-ID: test-case-009\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# valid case: no space between display name and URI in From header,
INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: Alice<sip:alice@there.com>;tag=1001\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74b1239\r\n\
Max-Forwards: 70\r\n\
To: Bob Jones<sip:bob@example.com>\r\n\
Call-ID: test-case-010\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# valid case: no space between header display name from,
INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From:Alice <sip:alice@there.com>;tag=1001\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74b1231239\r\n\
Max-Forwards: 70\r\n\
To:Bob Jones <sip:bob@example.com>\r\n\
Call-ID: test-case-011\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# valid case: no space between header display name and URI in From header,
INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From:Alice<sip:alice@there.com>;tag=1001\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4b123K74b1239\r\n\
Max-Forwards: 70\r\n\
To:Bob Jones<sip:bob@example.com>\r\n\
Call-ID: test-case-012\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# valid case: multiple space between header display name from,
INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From:     Alice    <sip:alice@there.com>;tag=1001\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74b1231239\r\n\
Max-Forwards: 70\r\n\
To:    Bob Jones                 <sip:bob@example.com>\r\n\
Call-ID: test-case-013\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"

# Invalid case: Empty angle brackets in From header
INVITE_MSG="INVITE sip:alice@example.com SIP/2.0\r\n\
From: <>;tag=1009\r\n\
Via: SIP/2.0/UDP there.com:5060;rport;branch=z9hG4bK74bf9\r\n\
Max-Forwards: 70\r\n\
To: <sip:bob@example.com>\r\n\
Call-ID: test-case-014\r\n\
CSeq: 1 INVITE\r\n\
Contact: <sip:alice@there.com>;expires=3600\r\n\
Content-Length: 0\r\n\r\n"
send_invite "$INVITE_MSG"
```
</details>

File: src/modules/uac/replace.c [hunk 2]
```diff
@@ -309,40 +309,85 @@ int replace_uri(struct sip_msg *msg, str *display, str *uri,
...
+				del_len++;
+				p--;
+			}
+			/* extend forward to include all spaces after display */
+			p = body->display.s + body->display.len;
+		} else {
+			/* no existing display - check spaces after colon
+			 * hdr->name is "To"/"From" without ':', skip past it */
+			del_offset = (hdr->name.s + hdr->name.len + 1) - msg->buf;
+			p = hdr->name.s + hdr->name.len + 1;
```
Comment (miconda) [inline src/modules/uac/replace.c]: I see that code is assuming that `:` is immediately after header name. The specs for HTTP1.1 forbid space between header name and `:` (https://datatracker.ietf.org/doc/html/rfc7230, section 3.2.4), but I couldn't find quickly a clear statement about it. I am not really confident that I remember exactly, but I think I saw in the past cases where SIP UAs put spaces there.

So I am making a note here, maybe other can come with better references.

In general SER/Kamailio tried to be flexible in accepting "unclear" variants. Probably a safety check is to see if the existing sip header parser in the code is ok with whitespaces in that place.

Comment (henningw) [reply src/modules/uac/replace.c]: Thanks for the feedback. Yes, I think its valid in SIP to have additional spaces between header name and ":". We spend some time thinking about it this afternoon. Its quite complicated and even the current PR version does not cover all "easy" cases.

Xenofon will probably fix the reported issue (additional second whitespace) first, possible also for the stable branches. And then we can investigate how to solve it in a more general way in git master. Maybe we should just keep it like it is. If somebody needs to have a complete header cleanup, this is probably more a job of a B2BUA.

Comment (miconda) [reply src/modules/uac/replace.c]: Probably it is indeed better to focus on the scope of the function to remove the display name, and if it is no new value for display name, I think the right logic is to remove from the start of the existing display name till the start of the uri field, not caring about what it is before the display name.

If one wants to `pretty-print` all header like `HName: hbody\r\n` can use headers iterator to reformat.

**Comment by miconda** [comment]
Thanks!
